### PR TITLE
 Pricing alert checks added.

### PR DESCRIPTION
**Pull request by saleem-latif**
This PR adds pricing alert for users who have used up available resources and need to upgrade account in order to continue using codesync.
Screenshot 2022-12-16 at 2 42 07 PM

### CODE REVIEW

**Review by Basit-qc — CHANGES_REQUESTED** [review]

File: src/main/java/org/intellij/sdk/codesync/utils/PricingAlerts.java [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+        acquirePricingLock();
+        boolean shouldUpgrade = CodeSyncMessages.showYesNoMessage(
+                Notification.UPGRADE, Notification.UPGRADE_PRICING_PLAN, CommonUtils.getCurrentProject()
+        );
+        if (shouldUpgrade) {
+            BrowserUtil.browse(CODESYNC_PRICING_URL);
+        }
+    }
+
+    public static void setPlanLimitReached(int repoId) {
```
[the remark this answers is not in the document]
Comment (saleem-latif) [reply src/main/java/org/intellij/sdk/codesync/utils/PricingAlerts.java:26]: Not, yet but I will add some usages a bit later.

File: src/main/java/org/intellij/sdk/codesync/utils/PricingAlerts.java [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+    public static void setPlanLimitReached() {
+        acquirePricingLock();
+        boolean shouldUpgrade = CodeSyncMessages.showYesNoMessage(
+                Notification.UPGRADE, Notification.UPGRADE_PRICING_PLAN, CommonUtils.getCurrentProject()
+        );
+        if (shouldUpgrade) {
+            BrowserUtil.browse(CODESYNC_PRICING_URL);
+        }
+    }
```
[the remark this answers is not in the document]
Comment (saleem-latif) [reply src/main/java/org/intellij/sdk/codesync/utils/PricingAlerts.java:24]: Yes, I have achieved that with expiring locks. The lock that I acquire expire after the specified time and then code retries the upload and sets the lock again if it fails.

File: src/main/java/org/intellij/sdk/codesync/clients/CodeSyncClient.java [hunk 3]
```diff
@@ -212,19 +199,32 @@ public void uploadToS3(File originalsFile, Map<String, Object> preSignedURLData)
...
-
             CodeSyncLogger.critical(String.format("Error while repo init, %s", error.getMessage()));
             return null;
         }
 
+        try {
+            jsonResponse.raiseForStatus();
+            return jsonResponse.getJsonResponse();
+        } catch (StatusCodeError statusCodeError) {
+            if (statusCodeError.getStatusCode()  == 402) {
```
[the remark this answers is not in the document]
Comment (saleem-latif) [reply src/main/java/org/intellij/sdk/codesync/clients/CodeSyncClient.java]: Done

File: src/main/java/org/intellij/sdk/codesync/clients/CodeSyncWebSocketClient.java [hunk 4]
```diff
@@ -171,7 +179,12 @@ public void sendDiffs(ArrayList<Pair<Integer, DiffFile>> diffsToSend, DataTransm
                 Long statusCode = (Long) response.get("status");
                 String diffFilePath = (String) response.get("diff_file_path");
                 if (statusCode != 200) {
-                    CodeSyncLogger.critical(String.format("Diff upload failed with error: %s.", response.get("error")));
+                    if (statusCode == Constants.ErrorCodes.DIFFS_LIMIT_REACHED) {
```
Comment (Basit-qc) [inline src/main/java/org/intellij/sdk/codesync/clients/CodeSyncWebSocketClient.java:182]: Authenticate does not return 402, diff-send returns 402.  I think this logic should go under `.sendDiffs`

**Review by Basit-qc — APPROVED** [review]

File: src/main/java/org/intellij/sdk/codesync/clients/CodeSyncWebSocketClient.java (hunk 4 above)
Comment (saleem-latif) [reply src/main/java/org/intellij/sdk/codesync/clients/CodeSyncWebSocketClient.java:182]: @Basit-qc I think the functionality is already in the `.sendDiffs`, authentication happens in another step and this is the status code returned by the diffs processing endpoint.

File: src/main/java/org/intellij/sdk/codesync/Constants.java [hunk 5]
```diff
@@ -47,28 +51,26 @@ private Constants() {
...
-    public static String WEBAPP_DASHBOARD_URL = String.format("%s/", WEB_APP_URL);
-    public static String PLANS_URL = String.format("%s/plans", WEB_APP_URL);
-    public static String REPO_PLAYBACK_LINK = String.format("%s/repos", WEB_APP_URL) + "/%s/playback";
-    public static String FILE_PLAYBACK_LINK = String.format("%s/files", WEB_APP_URL) + "/%s/history";
-
-    public static Integer DELAY_BETWEEN_BUFFER_TASKS = 5000;
-    public static String DATE_TIME_FORMAT = "yyyy-MM-dd HH:mm:ss.SSS z";
-    public static String DIFF_SOURCE = "intellij";
+    public static String WEBAPP_DASHBOARD_URL = String.format("%s/?%s", WEB_APP_URL, GA4_PARAMS);
+    public static String PLANS_URL = String.format("%s/plans?%s", WEB_APP_URL, GA4_PARAMS);
```
[the remark this answers is not in the document]
Comment (saleem-latif) [reply src/main/java/org/intellij/sdk/codesync/Constants.java:71]: 👍🏼 I will make this change as well.

File: src/main/java/org/intellij/sdk/codesync/clients/CodeSyncWebSocketClient.java (hunk 4 above)
Comment (Basit-qc) [reply src/main/java/org/intellij/sdk/codesync/clients/CodeSyncWebSocketClient.java:182]: Got it, thanks

**Pull request merged**